rhombus/rhombus/scribblings/ref-rx-pattern.scrbl

```diff
@@ -0,0 +1,664 @@
+  rx'"a" || "b"'.match("a")
+  rx'"a" || "b"'.match("b")
+  rx'"a" || "b"'.match("c")
+)
+
+}
+
+
+@doc(
+  rx.macro '#%parens ($pat)'
+){
+
+ A parenthesized pattern is equivalent to the @rhombus(pat) inside the
+ parentheses. That is, parentheses are just for grouping and resolving
+ precedence mismatches. See @rhombus($, ~at rhombus/rx) for inforation
+ about @tech{capture groups}, which are not implicitly created by
+ parentheses (as they are in some traditional regexp languages).
+
+@examples(
+  ~eval: rx_eval
+  rx'"a" || "b" ++ "c"'.match("ac")
+  rx'("a" || "b") ++ "c"'.match("ac")
+)
+
+}
+
+
+@doc(
+  rx.macro '#%brackets [$charset]'
+  rx.macro '$pat #%index [$charset]'
+){
+
+ A @brackets pattern, which is an implicit use of
+ @rhombus(#%brackets, ~at rhombus/rx), matches a single character or
+ byte, where @rhombus(charset) determines the matching characters or
+ bytes. An implicit @rhombus(#%index, ~at rhombus/rx) form (see
+ @secref("implicit")) is treated as a sequence of a @rhombus(pat) and
+ @rhombus(#%brackets, ~at rhombus/rx).
+
+ See @secref("rx-charset") for @tech{character set} forms that can be
+ used in @rhombus(charset).
+
+@examples(
+  ~eval: rx_eval
+  rx'["a"-"z"]'.match("m")
+  rx'["a"-"z"]'.match("0")
+)
+
+}
+
+
+@doc(
+  rx.macro '$pat *'
+  rx.macro '$pat * $mode'
+  grammar mode:
+    ~greedy
+    ~nongreedy
+    ~possessive
+){
+
+ Matches a sequence of 0 or more matches to @rhombus(pat).
+
+@examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'any*'.match("abc")
+    rx'any*'.match("")
+)
+
+ By default, the match uses @rhombus(~greedy) mode, where a larger number
+ of matches is tried first---but subsequent patterns may cause
+ backtracking to a shorter match. In @rhombus(~nongreedy) mode, shorter
+ matches are tried first. The @rhombus(~possessive) mode is like
+ @rhombus(~greedy), but without backtracking (i.e., the longest match
+ must succeed overall for the enclosing pattern); see also
+ @rhombus(cut, ~at rhombus/rx).
+
+@examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'($head: any*) ($tail: any*)'.match("abc")
+    rx'($head: any* ~nongreedy) ($tail: any*)'.match("abc")
+  ~repl:
+    rx'any* ~greedy "z"'.match("abcz")
+    rx'any* ~possessive "z"'.match("abcz")
+)
+
+}
+
+@doc(
+  ~nonterminal:
+    mode: * ~at rhombus/rx
+  rx.macro '$pat +'
+  rx.macro '$pat + $mode'
+){
+
+ Like @rhombus(*, ~at rhombus/rx), but matches 1 or more instances of @rhombus(pat).
+
+@examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'any+'.match("abc")
+    rx'any+'.match("")
+)
+
+}
+
+
+@doc(
+  ~nonterminal:
+    mode: * ~at rhombus/rx
+  rx.macro '$pat ?'
+  rx.macro '$pat ? $mode'
+){
+
+ Similar to @rhombus(*, ~at rhombus/rx), but matches 0 or 1 instances of @rhombus(pat).
+
+@examples(
+  ~eval: rx_eval
+  ~repl:
+    rx'any?'.match("a")
+    rx'any?'.match("")
+    rx'any?'.match("abc")
+)
+
+}
+
+
+@doc(
+  rx.macro '$pat #%comp {$count}'
+  rx.macro '$pat #%comp {$min ..}'
+  rx.macro '$pat #%comp {$min .. $max}'
```

Comment:
I added a margin note to point out that `.. max` as an expression creates a range that starts at `#neginf`, and the intent is to avoid suggesting negative counts are possible.  We could say that `.. max` means something different here or that negative values are treated as 0, but I thought it would be better to just avoid it.